File: sdks/dotnet/src/Rrflow.Rrd.Client/RrdClient.cs
using System;
using System.Buffers;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;

namespace Rrflow.Rrd;

public sealed partial class RrdClient : IDisposable
{
    private const int MaximumResponseBytes = 16 * 1024 * 1024;
    private static readonly HashSet<string> ResourceKinds = new(StringComparer.Ordinal)
    {
        "organization", "estate", "project", "instance", "node", "shard", "collection",
        "table", "record", "transaction", "snapshot", "backup", "operation",
    };
    private static readonly JsonSerializerOptions SerializerOptions = new(JsonSerializerDefaults.Web);

    private readonly Uri _baseUri;
    private readonly string _instance;
    private readonly TimeSpan _requestTimeout;
    private readonly int _maxAttempts;
    private readonly int _maxResponseBytes;
    private readonly HttpClient _http;
    private readonly bool _ownsHttp;

    public RrdClient(RrdClientOptions options, HttpMessageHandler? handler = null)
    {
        ArgumentNullException.ThrowIfNull(options);
        _baseUri = LoopbackUri(options.BaseUri);
        _instance = Canonical(options.Instance, "instance");
        if (options.RequestTimeout <= TimeSpan.Zero || options.RequestTimeout > TimeSpan.FromMinutes(5))
        {
            throw new RrdClientException("request timeout must be in (0, 5m]");
        }
        if (options.MaxAttempts is < 1 or > 8)
        {
            throw new RrdClientException("max attempts must be in 1..=8");
        }
        if (options.MaxResponseBytes is < 1 or > MaximumResponseBytes)
        {
            throw new RrdClientException("response limit must be in 1..=16777216 bytes");
        }
        _requestTimeout = options.RequestTimeout;
        _maxAttempts = options.MaxAttempts;
        _maxResponseBytes = options.MaxResponseBytes;
        _ownsHttp = true;
        _http = handler is null
            ? new HttpClient(new HttpClientHandler { AllowAutoRedirect = false })
            : new HttpClient(handler, disposeHandler: false);
    }

    public async Task<JsonElement> CapabilitiesAsync(CancellationToken cancellationToken = default)
    {
        JsonElement result = await CallAsync(
            OperationId.CapabilitiesRead, null, new RequestOptions(), cancellationToken)
            .ConfigureAwait(false);
        if (result.GetProperty("protocol").GetString() != "rrd"
            || result.GetProperty("protocol_version").GetInt32() != 1
            || result.GetProperty("instance").GetProperty("id").GetString() != _instance)
        {
            throw new RrdClientException("RRD capability protocol or instance identity differs");
        }
        return result;
    }

    public Task<JsonElement> EndpointCatalogueAsync(CancellationToken cancellationToken = default) =>
        CallAsync(OperationId.EndpointCatalogue, null, new RequestOptions(), cancellationToken);

    public Task<JsonElement> OpenApiAsync(CancellationToken cancellationToken = default) =>
        CallAsync(OperationId.OpenapiRead, null, new RequestOptions(), cancellationToken);

    public async Task<Session> CreateSessionAsync(
        string principalId,
        string credential,
        object payload,
        RequestOptions options,
        CancellationToken cancellationToken = default)
    {
        string principal = Canonical(principalId, "principal");
        if (string.IsNullOrEmpty(credential))
        {
            throw new RrdClientException("API-key credential must not be empty");
        }
        JsonElement result = await CallAsync(
            OperationId.SessionCreate,
            payload,
            options with { Session = null, ApiKey = new ApiKeyCredentials(principal, credential) },
            cancellationToken).ConfigureAwait(false);
        string sessionId = result.GetProperty("session_id").GetString() ?? string.Empty;
        string token = result.GetProperty("token").GetString() ?? string.Empty;
        if (sessionId.Length == 0 || token.Length == 0)
        {
            throw new RrdClientException("invalid RRD session lease identity");
        }
        return new Session(principal, new SessionLease(sessionId, token));
    }

    public async Task<JsonElement> CallAsync(
        OperationId operation,
        object? payload,
        RequestOptions options,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(options);
        Endpoint endpoint = EndpointCatalog.Get(operation);
        Dictionary<string, object?>? requestContext = MakeContext(endpoint, options);
        string path = ResolvePath(endpoint.Path, options.PathParameters);
        Uri target = new(_baseUri, path.TrimStart('/'));
        byte[]? body = null;
        if (endpoint.Method != "GET")
        {
            IReadOnlyList<ResourceSegment> resource = options.Resource
                ?? DefaultResource(options.PathParameters);
            body = JsonSerializer.SerializeToUtf8Bytes(new Dictionary<string, object?>
            {
                ["protocol"] = "rrd",
                ["protocol_version"] = 1,
                ["context"] = requestContext,
                ["resource"] = new Dictionary<string, object?>
                {
                    ["segments"] = ValidateResource(resource),
                },
                ["payload"] = payload ?? new Dictionary<string, object?>(),
            }, SerializerOptions);
        }
        bool retrySafe = endpoint.Method == "GET" || !endpoint.Mutation
            || options.IdempotencyKey is not null;
        int attempts = retrySafe ? _maxAttempts : 1;
        Exception? lastError = null;
        for (int attempt = 0; attempt < attempts; attempt++)
        {
            TimeSpan timeout = RemainingTimeout(options.Deadline);
            using CancellationTokenSource attemptTimeout = new(timeout);
            using CancellationTokenSource linked = CancellationTokenSource.CreateLinkedTokenSource(
                cancellationToken, attemptTimeout.Token);
            using HttpRequestMessage request = new(new HttpMethod(endpoint.Method), target);
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
            Authenticate(request, endpoint, options);
            if (body is not null)
            {
                request.Content = new ByteArrayContent(body);
                request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
            }
            try
            {
                using HttpResponseMessage response = await _http.SendAsync(
                    request, HttpCompletionOption.ResponseHeadersRead, linked.Token).ConfigureAwait(false);
                byte[] encoded = await ReadBoundedAsync(response, linked.Token).ConfigureAwait(false);
                return DecodeResponse((int)response.StatusCode, encoded, requestContext);
            }
            catch (Exception error) when (
                error is HttpRequestException
                || (error is OperationCanceledException && !cancellationToken.IsCancellationRequested))
            {
                lastError = error;
            }
        }
        cancellationToken.ThrowIfCancellationRequested();
        throw new RrdClientException("RRD transport failed", lastError!);
    }

    public void Dispose()
    {
        if (_ownsHttp)
        {
            _http.Dispose();
        }
    }

    private static Dictionary<string, object?>? MakeContext(Endpoint endpoint, RequestOptions options)
    {
        if (endpoint.Method == "GET")
        {
            return null;
        }
        Dictionary<string, object?> context = new()
        {
            ["request_id"] = Correlation(options.RequestId, "request ID"),
            ["operation_id"] = Correlation(options.OperationId, "operation ID"),
        };
        if (options.IdempotencyKey is not null)
        {
            context["idempotency_key"] = Correlation(options.IdempotencyKey, "idempotency key");
        }
        else if (endpoint.Mutation)
        {
            throw new RrdClientException("mutating requests require an idempotency key");
        }
        if (options.Deadline is not null)
        {
            long milliseconds = options.Deadline.Value.ToUnixTimeMilliseconds();
            if (milliseconds <= 0)
            {
                throw new RrdClientException("deadline must be a positive Unix millisecond instant");
            }
            context["deadline_unix_ms"] = milliseconds;
        }
        return context;
    }

    private static void Authenticate(HttpRequestMessage request, Endpoint endpoint, RequestOptions options)
    {
        if (endpoint.Authentication == Authentication.ApiKey)
        {
            if (options.ApiKey is null || string.IsNullOrEmpty(options.ApiKey.Credential))
            {
                throw new RrdClientException($"{endpoint.WireName} requires API-key authentication");
            }
            request.Headers.Add("X-RRD-Principal", Canonical(options.ApiKey.PrincipalId, "principal"));
            request.Headers.TryAddWithoutValidation("Authorization", "ApiKey " + options.ApiKey.Credential);
        }
        else if (endpoint.Authentication == Authentication.SessionBearer)
        {
            if (options.Session is null
                || options.Session.Lease.SessionId.Length == 0
                || options.Session.Lease.Token.Length == 0)
            {
                throw new RrdClientException($"{endpoint.WireName} requires a valid session");
            }
            request.Headers.Add("X-RRD-Session", options.Session.Lease.SessionId);
            request.Headers.TryAddWithoutValidation(
                "Authorization", "Bearer " + options.Session.Lease.Token);
        }
    }

    private static JsonElement DecodeResponse(
        int status,
        byte[] encoded,
        IReadOnlyDictionary<string, object?>? requestContext)
    {
        try
        {
            using JsonDocument document = JsonDocument.Parse(encoded);
            JsonElement envelope = document.RootElement;
            RequireFields(envelope, "protocol", "protocol_version", "request_id", "operation_id", "outcome");
            if (envelope.GetProperty("protocol").GetString() != "rrd"
                || envelope.GetProperty("protocol_version").GetInt32() != 1)
            {
                throw new RrdClientException("RRD response protocol differs");
            }
            if (requestContext is not null
                && (envelope.GetProperty("request_id").GetString()
                        != (string)requestContext["request_id"]!
                    || envelope.GetProperty("operation_id").GetString()
                        != (string)requestContext["operation_id"]!))
            {
                throw new RrdClientException("RRD response request/operation identity differs");
            }
            JsonElement outcome = envelope.GetProperty("outcome");
            string outcomeStatus = outcome.GetProperty("status").GetString() ?? string.Empty;
            bool success = status is >= 200 and < 300;
            if (success != (outcomeStatus == "ok"))
            {
                throw new RrdClientException("RRD HTTP status and typed outcome disagree");
            }
            if (outcomeStatus == "error")
            {
                RequireFields(outcome, "status", "error");
                JsonElement error = outcome.GetProperty("error");
                RequireFields(error, "code", "message", "retryable", "details");
                Dictionary<string, string> details = error.GetProperty("details")
                    .EnumerateObject().ToDictionary(
                        property => property.Name,
                        property => property.Value.GetString() ?? string.Empty,
                        StringComparer.Ordinal);
                throw new RrdApiException(
                    status,
                    error.GetProperty("code").GetString() ?? string.Empty,
                    error.GetProperty("message").GetString() ?? string.Empty,
                    error.GetProperty("retryable").GetBoolean(),
                    details);
            }
            if (outcomeStatus != "ok")
            {
                throw new RrdClientException("RRD response outcome status is invalid");
            }
            RequireFields(outcome, "status", "payload");
            JsonElement result = outcome.GetProperty("payload");
            if (result.ValueKind != JsonValueKind.Object)
            {
                throw new RrdClientException("RRD success payload must be an object");
            }
            return result.Clone();
        }
        catch (RrdClientException)
        {
            throw;
        }
        catch (Exception error) when (error is JsonException or InvalidOperationException)
        {
            throw new RrdClientException("RRD response envelope is invalid", error);
        }
    }

    private async Task<byte[]> ReadBoundedAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        if (response.Content.Headers.ContentLength > _maxResponseBytes)
        {
            throw new RrdClientException("RRD response exceeded the configured byte limit");
        }
        await using Stream input = await response.Content.ReadAsStreamAsync(cancellationToken)
            .ConfigureAwait(false);
        using MemoryStream output = new();
        byte[] buffer = ArrayPool<byte>.Shared.Rent(8192);
        try
        {
            int total = 0;
            int count;
            while ((count = await input.ReadAsync(buffer, cancellationToken).ConfigureAwait(false)) > 0)
            {
                total += count;
                if (total > _maxResponseBytes)
                {
                    throw new RrdClientException("RRD response exceeded the configured byte limit");
                }
                output.Write(buffer, 0, count);
            }
            return output.ToArray();
        }
        finally
        {
            ArrayPool<byte>.Shared.Return(buffer);
        }
    }

    private TimeSpan RemainingTimeout(DateTimeOffset? deadline)
    {
        if (deadline is null)
        {
            return _requestTimeout;
        }
        TimeSpan remaining = deadline.Value - DateTimeOffset.UtcNow;
        if (remaining <= TimeSpan.Zero)
        {
            throw new RrdClientException("RRD request deadline has expired");
        }
        return remaining < _requestTimeout ? remaining : _requestTimeout;
    }

    private IReadOnlyList<ResourceSegment> DefaultResource(
        IReadOnlyDictionary<string, string> parameters)
    {
        List<ResourceSegment> result = new();
        if (parameters.TryGetValue("estate", out string? estate))
        {
            result.Add(new ResourceSegment("estate", Canonical(estate, "estate")));
        }
        result.Add(new ResourceSegment("instance", _instance));
        return result;
    }

    private static IReadOnlyList<ResourceSegment> ValidateResource(
        IReadOnlyList<ResourceSegment> segments)
    {
        if (segments.Count is < 1 or > 16)
        {
            throw new RrdClientException("resource paths must contain 1..=16 segments");
        }
        HashSet<string> kinds = new(StringComparer.Ordinal);
        return segments.Select(segment =>
        {
            if (!ResourceKinds.Contains(segment.Kind) || !kinds.Add(segment.Kind))
            {
                throw new RrdClientException("resource kind is unknown or repeated");
            }
            return new ResourceSegment(segment.Kind, Canonical(segment.Id, "resource"));
        }).ToArray();
    }

    private static string ResolvePath(string template, IReadOnlyDictionary<string, string> parameters) =>
        PathParameter().Replace(template, match =>
        {
            string name = match.Groups[1].Value;
            if (!parameters.TryGetValue(name, out string? value) || value.Length == 0)
            {
                throw new RrdClientException($"missing path parameter {name}");
            }
            return Uri.EscapeDataString(Correlation(value, name + " path parameter"));
        });

    private static string Correlation(string? value, string label)
    {
        if (value is null || value.Length is < 1 or > 128 || !CorrelationPattern().IsMatch(value))
        {
            throw new RrdClientException($"{label} is not a canonical RRD correlation ID");
        }
        return value;
    }

    private static string Canonical(string? value, string label)
    {
        if (value is null || value.Length is < 1 or > 128 || !CanonicalPattern().IsMatch(value))
        {
            throw new RrdClientException($"{label} is not a canonical RRD identifier");
        }
        return value;
    }

    private static Uri LoopbackUri(Uri uri)
    {
        bool loopback = uri.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase)
            || (IPAddress.TryParse(uri.Host, out IPAddress? address) && IPAddress.IsLoopback(address));
        if (uri.Scheme != Uri.UriSchemeHttp
            || !loopback
            || !string.IsNullOrEmpty(uri.UserInfo)
            || !string.IsNullOrEmpty(uri.Query)
            || !string.IsNullOrEmpty(uri.Fragment))
        {
            throw new RrdClientException(
                "RRD .NET client permits only credential-free loopback HTTP before TLS qualification");
        }
        return new Uri(uri.AbsoluteUri.TrimEnd('/') + "/", UriKind.Absolute);
    }

    private static void RequireFields(JsonElement element, params string[] expected)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            throw new RrdClientException("RRD response object is invalid");
        }
        HashSet<string> actual = element.EnumerateObject()
            .Select(property => property.Name)
            .ToHashSet(StringComparer.Ordinal);
        if (!actual.SetEquals(expected))
        {
            throw new RrdClientException("RRD response object fields differ");
        }
    }

    [GeneratedRegex("^[A-Za-z0-9._:-]+$", RegexOptions.CultureInvariant)]
    private static partial Regex CorrelationPattern();

    [GeneratedRegex("^[a-z0-9][a-z0-9._-]*$", RegexOptions.CultureInvariant)]
    private static partial Regex CanonicalPattern();

    [GeneratedRegex("\\{([a-z]+)\\}", RegexOptions.CultureInvariant)]
    private static partial Regex PathParameter();
}

File: sdks/java/src/main/java/io/rrflow/rrd/RrdClient.java
package io.rrflow.rrd;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.InetAddress;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;

public final class RrdClient {
    private static final Pattern CORRELATION = Pattern.compile("^[A-Za-z0-9._:-]+$");
    private static final Pattern CANONICAL = Pattern.compile("^[a-z0-9][a-z0-9._-]*$");
    private static final Pattern PATH_PARAMETER = Pattern.compile("\\{([a-z]+)}");
    private static final Set<String> RESOURCE_KINDS = Set.of(
            "organization", "estate", "project", "instance", "node", "shard",
            "collection", "table", "record", "transaction", "snapshot", "backup",
            "operation");
    private static final int DEFAULT_RESPONSE_LIMIT = 4 * 1024 * 1024;
    private static final int MAXIMUM_RESPONSE_LIMIT = 16 * 1024 * 1024;
    private static final JsonMapper JSON = JsonMapper.builder().build();

    private final URI baseUri;
    private final String instance;
    private final Duration requestTimeout;
    private final int maxAttempts;
    private final int maxResponseBytes;
    private final HttpClient http;

    public RrdClient(String baseUrl, String instance) {
        this(baseUrl, instance, Duration.ofSeconds(5), 2, DEFAULT_RESPONSE_LIMIT, null);
    }

    public RrdClient(
            String baseUrl,
            String instance,
            Duration requestTimeout,
            int maxAttempts,
            int maxResponseBytes,
            HttpClient http) {
        this.baseUri = loopbackUri(baseUrl);
        this.instance = canonical(instance, "instance");
        if (requestTimeout.isZero() || requestTimeout.isNegative()
                || requestTimeout.compareTo(Duration.ofMinutes(5)) > 0) {
            throw new RrdClientException("request timeout must be in (0, 5m]");
        }
        if (maxAttempts < 1 || maxAttempts > 8) {
            throw new RrdClientException("max attempts must be in 1..=8");
        }
        if (maxResponseBytes < 1 || maxResponseBytes > MAXIMUM_RESPONSE_LIMIT) {
            throw new RrdClientException("response limit must be in 1..=16777216 bytes");
        }
        this.requestTimeout = requestTimeout;
        this.maxAttempts = maxAttempts;
        this.maxResponseBytes = maxResponseBytes;
        this.http = http == null
                ? HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER).build()
                : http;
    }

    public JsonNode capabilities() {
        JsonNode result = call(OperationId.CAPABILITIES_READ, null, RequestOptions.empty());
        if (!result.path("protocol").asString().equals("rrd")
                || result.path("protocol_version").asInt() != 1
                || !result.path("instance").path("id").asString().equals(instance)) {
            throw new RrdClientException("RRD capability protocol or instance identity differs");
        }
        return result;
    }

    public JsonNode endpointCatalogue() {
        return call(OperationId.ENDPOINT_CATALOGUE, null, RequestOptions.empty());
    }

    public JsonNode openApi() {
        return call(OperationId.OPENAPI_READ, null, RequestOptions.empty());
    }

    public Session createSession(
            String principalId, String credential, Object payload, RequestOptions options) {
        String principal = canonical(principalId, "principal");
        if (credential == null || credential.isEmpty()) {
            throw new RrdClientException("API-key credential must not be empty");
        }
        JsonNode result = call(
                OperationId.SESSION_CREATE, payload, options.withApiKey(principal, credential));
        String sessionId = result.path("session_id").asString();
        String token = result.path("token").asString();
        if (sessionId.isEmpty() || token.isEmpty()) {
            throw new RrdClientException("invalid RRD session lease identity");
        }
        return new Session(principal, new Session.SessionLease(sessionId, token));
    }

    public JsonNode call(OperationId operation, Object payload, RequestOptions options) {
        if (operation == null || options == null) {
            throw new RrdClientException("operation and request options are required");
        }
        Map<String, Object> context = requestContext(operation, options);
        String path = resolvePath(operation.path(), options.pathParameters());
        URI target = baseUri.resolve(path.startsWith("/") ? path.substring(1) : path);
        byte[] body = null;
        if (!operation.method().equals("GET")) {
            List<ResourceSegment> resource = options.resource();
            if (resource == null) {
                resource = defaultResource(options.pathParameters());
            }
            Map<String, Object> envelope = new HashMap<>();
            envelope.put("protocol", "rrd");
            envelope.put("protocol_version", 1);
            envelope.put("context", context);
            envelope.put("resource", Map.of("segments", validateResource(resource)));
            envelope.put("payload", payload == null ? Map.of() : payload);
            try {
                body = JSON.writeValueAsBytes(envelope);
            } catch (Exception error) {
                throw new RrdClientException("encode RRD request", error);
            }
        }
        boolean retrySafe = operation.method().equals("GET")
                || !operation.mutation()
                || options.idempotencyKey() != null;
        int attempts = retrySafe ? maxAttempts : 1;
        IOException lastError = null;
        for (int attempt = 0; attempt < attempts; attempt++) {
            Duration timeout = remainingTimeout(options.deadline());
            HttpRequest.Builder builder = HttpRequest.newBuilder(target)
                    .timeout(timeout)
                    .header("Accept", "application/json");
            authenticate(builder, operation, options);
            if (body == null) {
                builder.method(operation.method(), HttpRequest.BodyPublishers.noBody());
            } else {
                builder.header("Content-Type", "application/json")
                        .method(operation.method(), HttpRequest.BodyPublishers.ofByteArray(body));
            }
            try {
                HttpResponse<InputStream> response =
                        http.send(builder.build(), HttpResponse.BodyHandlers.ofInputStream());
                byte[] encoded = readBounded(response);
                return decodeResponse(response.statusCode(), encoded, context);
            } catch (IOException error) {
                lastError = error;
            } catch (InterruptedException error) {
                Thread.currentThread().interrupt();
                throw new RrdClientException("RRD request interrupted", error);
            }
        }
        throw new RrdClientException("RRD transport failed", lastError);
    }

    private void authenticate(
            HttpRequest.Builder builder, OperationId operation, RequestOptions options) {
        switch (operation.authentication()) {
            case PUBLIC -> { }
            case API_KEY -> {
                RequestOptions.ApiKey apiKey = options.apiKey();
                if (apiKey == null || apiKey.credential() == null || apiKey.credential().isEmpty()) {
                    throw new RrdClientException(operation.wireName() + " requires API-key authentication");
                }
                builder.header("X-RRD-Principal", canonical(apiKey.principalId(), "principal"));
                builder.header("Authorization", "ApiKey " + apiKey.credential());
            }
            case SESSION_BEARER -> {
                Session session = options.session();
                if (session == null || session.lease().sessionId().isEmpty()
                        || session.lease().token().isEmpty()) {
                    throw new RrdClientException(operation.wireName() + " requires a valid session");
                }
                builder.header("X-RRD-Session", session.lease().sessionId());
                builder.header("Authorization", "Bearer " + session.lease().token());
            }
        }
    }

    private Map<String, Object> requestContext(OperationId operation, RequestOptions options) {
        if (operation.method().equals("GET")) {
            return null;
        }
        Map<String, Object> context = new HashMap<>();
        context.put("request_id", correlation(options.requestId(), "request ID"));
        context.put("operation_id", correlation(options.operationId(), "operation ID"));
        if (options.idempotencyKey() != null) {
            context.put("idempotency_key", correlation(options.idempotencyKey(), "idempotency key"));
        } else if (operation.mutation()) {
            throw new RrdClientException("mutating requests require an idempotency key");
        }
        if (options.deadline() != null) {
            long milliseconds = options.deadline().toEpochMilli();
            if (milliseconds <= 0) {
                throw new RrdClientException("deadline must be a positive Unix millisecond instant");
            }
            context.put("deadline_unix_ms", milliseconds);
        }
        return context;
    }

    private JsonNode decodeResponse(int status, byte[] encoded, Map<String, Object> context) {
        final JsonNode envelope;
        try {
            envelope = JSON.readTree(encoded);
        } catch (Exception error) {
            throw new RrdClientException("RRD response envelope is invalid", error);
        }
        requireExactFields(envelope, Set.of(
                "protocol", "protocol_version", "request_id", "operation_id", "outcome"));
        if (!envelope.path("protocol").asString().equals("rrd")
                || envelope.path("protocol_version").asInt() != 1) {
            throw new RrdClientException("RRD response protocol differs");
        }
        if (context != null && (!envelope.path("request_id").asString().equals(context.get("request_id"))
                || !envelope.path("operation_id").asString().equals(context.get("operation_id")))) {
            throw new RrdClientException("RRD response request/operation identity differs");
        }
        JsonNode outcome = envelope.path("outcome");
        String outcomeStatus = outcome.path("status").asString();
        boolean success = status >= 200 && status < 300;
        if (success != outcomeStatus.equals("ok")) {
            throw new RrdClientException("RRD HTTP status and typed outcome disagree");
        }
        if (outcomeStatus.equals("error")) {
            requireExactFields(outcome, Set.of("status", "error"));
            JsonNode error = outcome.path("error");
            requireExactFields(error, Set.of("code", "message", "retryable", "details"));
            if (error.path("code").asString().isEmpty()
                    || error.path("message").asString().isEmpty()
                    || !error.path("retryable").isBoolean()
                    || !error.path("details").isObject()) {
                throw new RrdClientException("RRD error outcome is incomplete");
            }
            Map<String, String> details = new HashMap<>();
            error.path("details").properties().forEach(entry ->
                    details.put(entry.getKey(), entry.getValue().asString()));
            throw new RrdApiException(
                    status, error.path("code").asString(), error.path("message").asString(),
                    error.path("retryable").asBoolean(), details);
        }
        if (!outcomeStatus.equals("ok")) {
            throw new RrdClientException("RRD response outcome status is invalid");
        }
        requireExactFields(outcome, Set.of("status", "payload"));
        JsonNode result = outcome.path("payload");
        if (!result.isObject()) {
            throw new RrdClientException("RRD success payload must be an object");
        }
        return result;
    }

    private static void requireExactFields(JsonNode node, Set<String> expected) {
        if (node == null || !node.isObject()) {
            throw new RrdClientException("RRD response object is invalid");
        }
        Set<String> actual = new HashSet<>();
        actual.addAll(node.propertyNames());
        if (!actual.equals(expected)) {
            throw new RrdClientException("RRD response object fields differ");
        }
    }

    private byte[] readBounded(HttpResponse<InputStream> response) throws IOException {
        long declared = response.headers().firstValueAsLong("Content-Length").orElse(-1);
        if (declared > maxResponseBytes) {
            throw new RrdClientException("RRD response exceeded the configured byte limit");
        }
        try (InputStream input = response.body(); ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[8192];
            int total = 0;
            int count;
            while ((count = input.read(buffer)) != -1) {
                total += count;
                if (total > maxResponseBytes) {
                    throw new RrdClientException("RRD response exceeded the configured byte limit");
                }
                output.write(buffer, 0, count);
            }
            return output.toByteArray();
        }
    }

    private Duration remainingTimeout(Instant deadline) {
        if (deadline == null) {
            return requestTimeout;
        }
        Duration remaining = Duration.between(Instant.now(), deadline);
        if (remaining.isZero() || remaining.isNegative()) {
            throw new RrdClientException("RRD request deadline has expired");
        }
        return remaining.compareTo(requestTimeout) < 0 ? remaining : requestTimeout;
    }

    private List<ResourceSegment> defaultResource(Map<String, String> parameters) {
        List<ResourceSegment> result = new ArrayList<>();
        String estate = parameters.get("estate");
        if (estate != null) {
            result.add(new ResourceSegment("estate", canonical(estate, "estate")));
        }
        result.add(new ResourceSegment("instance", instance));
        return result;
    }

    private static List<ResourceSegment> validateResource(List<ResourceSegment> segments) {
        if (segments.isEmpty() || segments.size() > 16) {
            throw new RrdClientException("resource paths must contain 1..=16 segments");
        }
        Set<String> kinds = new HashSet<>();
        List<ResourceSegment> result = new ArrayList<>();
        for (ResourceSegment segment : segments) {
            if (!RESOURCE_KINDS.contains(segment.kind()) || !kinds.add(segment.kind())) {
                throw new RrdClientException("resource kind is unknown or repeated");
            }
            result.add(new ResourceSegment(segment.kind(), canonical(segment.id(), "resource")));
        }
        return List.copyOf(result);
    }

    private static String resolvePath(String template, Map<String, String> parameters) {
        Matcher matcher = PATH_PARAMETER.matcher(template);
        StringBuilder result = new StringBuilder();
        while (matcher.find()) {
            String value = parameters.get(matcher.group(1));
            if (value == null || value.isEmpty()) {
                throw new RrdClientException("missing path parameter " + matcher.group(1));
            }
            String encoded = URLEncoder.encode(
                    correlation(value, matcher.group(1) + " path parameter"), StandardCharsets.UTF_8)
                    .replace("+", "%20");
            matcher.appendReplacement(result, Matcher.quoteReplacement(encoded));
        }
        matcher.appendTail(result);
        return result.toString();
    }

    private static String correlation(String value, String label) {
        if (value == null || value.length() > 128 || !CORRELATION.matcher(value).matches()) {
            throw new RrdClientException(label + " is not a canonical RRD correlation ID");
        }
        return value;
    }

    private static String canonical(String value, String label) {
        if (value == null || value.length() > 128 || !CANONICAL.matcher(value).matches()) {
            throw new RrdClientException(label + " is not a canonical RRD identifier");
        }
        return value;
    }

    private static URI loopbackUri(String value) {
        final URI uri;
        try {
            uri = URI.create(value);
            String host = uri.getHost();
            boolean loopback = "localhost".equals(host);
            if (host != null && !loopback && host.matches("^[0-9a-fA-F:.]+$")) {
                InetAddress address = InetAddress.getByName(host);
                loopback = address.isLoopbackAddress();
            }
            if (!"http".equals(uri.getScheme()) || !loopback || uri.getUserInfo() != null
                    || uri.getQuery() != null || uri.getFragment() != null) {
                throw new IllegalArgumentException();
            }
            return URI.create(value.endsWith("/") ? value : value + "/");
        } catch (Exception error) {
            throw new RrdClientException(
                    "RRD Java client permits only credential-free loopback HTTP before TLS qualification",
                    error);
        }
    }
}
